### PR TITLE
[clang-format] Option to ignore macro definitions

### PR DESCRIPTION
**Pull request by tomekpaszek**
Sometimes macro definitions contain a layout we don't want to change in any context. This PR adds a new style option `IgnorePPDefinitions` that prevents clang-format from touching macro definitions.

This PR addresses issue #67991

### CODE REVIEW

File: clang/include/clang/Format/Format.h [hunk 1]
```diff
@@ -4648,6 +4648,10 @@ struct FormatStyle {
   /// \version 11
   std::vector<std::string> WhitespaceSensitiveMacros;
 
+  /// Ignore formatting in preprocessor definitions.
+  /// \version 18
+  bool IgnorePPDefinitions;
```
Comment (HazardyKnusperkeks) [inline clang/include/clang/Format/Format.h:4653]: Please sort alphabetically.

File: clang/lib/Format/UnwrappedLineFormatter.cpp [hunk 2]
```diff
@@ -1354,7 +1354,9 @@ unsigned UnwrappedLineFormatter::format(
 
     bool FixIndentation = (FixBadIndentation || ContinueFormatting) &&
                           Indent != TheLine.First->OriginalColumn;
-    bool ShouldFormat = TheLine.Affected || FixIndentation;
+    bool ShouldFormat = (TheLine.Affected || FixIndentation);
```
Comment (HazardyKnusperkeks) [inline clang/lib/Format/UnwrappedLineFormatter.cpp:1357]: Why adding the parenthesis? (Actually clang-format, if new enough, should remove them.)

File: clang/unittests/Format/FormatTest.cpp [hunk 3]
```diff
@@ -24153,6 +24153,23 @@ TEST_F(FormatTest, WhitespaceSensitiveMacros) {
   verifyNoChange("FOO(String-ized&Messy+But: :Still=Intentional);", Style);
 }
 
+TEST_F(FormatTest, IgnorePPDefinitions) {
```
Comment (HazardyKnusperkeks) [inline clang/unittests/Format/FormatTest.cpp:24156]: Add some tests:
* With formatted code above and below.
* Breaking the column limit.
* Multiline definitions (with different positions for the `\`).

We need to see how you think it should be formatted.

**Review by HazardyKnusperkeks — CHANGES_REQUESTED** [review]

**Comment by rymiel** [comment]
https://github.com/llvm/llvm-project/pull/67911 is a pull request, not an issue, i'm not sure what you're referring to

File: clang/include/clang/Format/Format.h (hunk 1 above)
Comment (rymiel) [reply clang/include/clang/Format/Format.h:4653]: (this applies to not just here, but below on line 4818, and in Format.cpp, and in ConfigParseTest.cpp)

**Review by owenca — CHANGES_REQUESTED** [review]
Please refer to the [policy](https://clang.llvm.org/docs/ClangFormatStyleOptions.html#adding-additional-style-options) for adding new options.

**Comment by owenca** [comment]
> #67911 is a pull request, not an issue, i'm not sure what you're referring to

I think it was a typo. It should be #67991.

**Review by mydeveloperday — CHANGES_REQUESTED** [review]

**Comment by tru** [comment]
@tomekpaszek do you plan to finish this PR? we have been testing it internally on our code and would really like this to land in upstream LLVM.

**Comment by tomekpaszek** [comment]
Yes, I just couldn't find time for it. I'll try to look into it within the next few days. Great to hear you there is interest in using it!

File: clang/include/clang/Format/Format.h (hunk 1 above)
Comment (tomekpaszek) [reply clang/include/clang/Format/Format.h:4657]: fixed

File: clang/lib/Format/UnwrappedLineFormatter.cpp (hunk 2 above)
Comment (tomekpaszek) [reply clang/lib/Format/UnwrappedLineFormatter.cpp]: this was unintentional, fixed now

File: clang/unittests/Format/FormatTest.cpp (hunk 3 above)
Comment (tomekpaszek) [reply clang/unittests/Format/FormatTest.cpp:24156]: added more tests now

File: clang/unittests/Format/FormatTest.cpp [hunk 4]
```diff
@@ -24153,6 +24153,23 @@ TEST_F(FormatTest, WhitespaceSensitiveMacros) {
...
+  verifyNoChange("#define  A", Style);
+  verifyNoChange("#define A   b", Style);
+  verifyNoChange("#define A  (  args   )", Style);
+  verifyNoChange("#define A  (  args   )  =  func  (  args  )", Style);
+  verifyNoChange("#define TEXT \\\nLine  number  one .  \\\nNumber  two .",
+                 Style);
+  verifyNoChange("#define A x:", Style);
+  verifyNoChange("#define A a. b", Style);
+
+  // TODO
```
[the remark this answers is not in the document]
Comment (tomekpaszek) [reply clang/unittests/Format/FormatTest.cpp]: The code path for when the statement is preceded by a comment seems different. It's not 100% unclear to me where to look to support that. Any hints would be appriciated

File: clang/include/clang/Format/Format.h [hunk 5]
```diff
@@ -4719,6 +4723,7 @@ struct FormatStyle {
                R.IncludeStyle.IncludeIsMainRegex &&
            IncludeStyle.IncludeIsMainSourceRegex ==
                R.IncludeStyle.IncludeIsMainSourceRegex &&
+           IgnorePPDefinitions == R.IgnorePPDefinitions &&
```
Comment (HazardyKnusperkeks) [inline clang/include/clang/Format/Format.h:4726]: A few lines up.

File: clang/unittests/Format/FormatTest.cpp [hunk 6]
```diff
@@ -24153,6 +24153,53 @@ TEST_F(FormatTest, WhitespaceSensitiveMacros) {
...
+
+  verifyNoChange("#define  A", Style);
+  verifyNoChange("#define A   b", Style);
+  verifyNoChange("#define A  (  args   )", Style);
+  verifyNoChange("#define A  (  args   )  =  func  (  args  )", Style);
+
+  verifyNoChange("#define A x:", Style);
+  verifyNoChange("#define A a. b", Style);
+    
+  // Surrounded with formatted code
```
Comment (HazardyKnusperkeks) [inline clang/unittests/Format/FormatTest.cpp:24168]:
```suggestion
  // Surrounded with formatted code.
```
Always end comments on full stop.

File: clang/unittests/Format/FormatTest.cpp [hunk 7]
```diff
@@ -24153,6 +24153,53 @@ TEST_F(FormatTest, WhitespaceSensitiveMacros) {
...
+  Style.ColumnLimit = 0;
+    
+  // Multiline definition
+  verifyNoChange("#define A \\\n"
+                 "Line one with spaces  .  \\\n"
+                 " Line two.",
+                   Style);
+  verifyNoChange("#define A \\\n"
+                 "a a \\\n"
+                 "a        \\\na",
```
Comment (HazardyKnusperkeks) [inline clang/unittests/Format/FormatTest.cpp:24189]:
```suggestion
                 "a        \\\n"
                 "a",
```

For easier reading.

File: clang/unittests/Format/FormatTest.cpp (hunk 3 above)
Comment (HazardyKnusperkeks) [reply clang/unittests/Format/FormatTest.cpp:24156]: Can you also add some nested `#if` and `#define` to show the (lacking?) indentation?

File: clang/lib/Format/UnwrappedLineParser.cpp [hunk 8]
```diff
@@ -1134,6 +1134,14 @@ void UnwrappedLineParser::parsePPDefine() {
     return;
   }
 
+  if (Style.IgnorePPDefinitions) {
+    do {
+      nextToken();
+    } while (!eof());
```
Comment (HazardyKnusperkeks) [inline clang/lib/Format/UnwrappedLineParser.cpp:1140]: I assume you have tested this on real code. Can you explain why it is okay to check for `eof()`? I really don't know this code, and am a bit surprised.

File: clang/unittests/Format/FormatTest.cpp (hunk 4 above)
[the remark this answers is not in the document]
Comment (HazardyKnusperkeks) [reply clang/unittests/Format/FormatTest.cpp]: I always comment every other check out, and run only this test case in a debugger to look where the change is needed.

[the remark this answers is not in the document]
Comment (HazardyKnusperkeks) [reply clang/unittests/Format/FormatTest.cpp]: Actually tests with a (line and block) comment after the `#define` would be interesting too, should they be wrapped? Or the space after `//` added?

File: clang/lib/Format/UnwrappedLineFormatter.cpp [hunk 9]
```diff
@@ -1355,6 +1355,8 @@ unsigned UnwrappedLineFormatter::format(
     bool FixIndentation = (FixBadIndentation || ContinueFormatting) &&
                           Indent != TheLine.First->OriginalColumn;
     bool ShouldFormat = TheLine.Affected || FixIndentation;
+    if (Style.IgnorePPDefinitions && TheLine.Type == LT_PreprocessorDirective)
```
Comment (HazardyKnusperkeks) [inline clang/lib/Format/UnwrappedLineFormatter.cpp:1358]: Wouldn't this take all preprocessor directives from the formatting? This is at least not what the documentation states.

File: clang/include/clang/Format/Format.h [hunk 10]
```diff
@@ -4648,6 +4648,10 @@ struct FormatStyle {
   /// \version 11
   std::vector<std::string> WhitespaceSensitiveMacros;
 
+  /// Ignore formatting in preprocessor definitions.
```
[the remark this answers is not in the document]
Comment (HazardyKnusperkeks) [reply clang/include/clang/Format/Format.h]: Run `clang/docs/tools/dump_format_style.py`.

File: clang/unittests/Format/FormatTest.cpp [hunk 11]
```diff
@@ -24153,6 +24153,53 @@ TEST_F(FormatTest, WhitespaceSensitiveMacros) {
...
+                 Style);
```
Comment (HazardyKnusperkeks) [inline clang/unittests/Format/FormatTest.cpp:24190]:
```suggestion
                Style);
```

Comment (HazardyKnusperkeks) [reply clang/unittests/Format/FormatTest.cpp:24190]: This doesn't look formatted.

File: clang/unittests/Format/FormatTest.cpp (hunk 3 above)
Comment (tomekpaszek) [reply clang/unittests/Format/FormatTest.cpp:24156]: Good idea. Added cases to cover that

File: clang/include/clang/Format/Format.h (hunk 10 above)
[the remark this answers is not in the document]
Comment (tomekpaszek) [reply clang/include/clang/Format/Format.h]: thanks for the tip!

File: clang/lib/Format/UnwrappedLineFormatter.cpp [hunk 12]
```diff
@@ -1355,8 +1355,11 @@ unsigned UnwrappedLineFormatter::format(
     bool FixIndentation = (FixBadIndentation || ContinueFormatting) &&
                           Indent != TheLine.First->OriginalColumn;
     bool ShouldFormat = TheLine.Affected || FixIndentation;
-    if (Style.IgnorePPDefinitions && TheLine.Type == LT_PreprocessorDirective)
+    if (Style.IgnorePPDefinitions && TheLine.Type == LT_PreprocessorDirective &&
+        TheLine.getFirstNonComment()->Next->is(tok::pp_define)) {
```
Comment (HazardyKnusperkeks) [inline clang/lib/Format/UnwrappedLineFormatter.cpp:1359]: Does `Next` always exist? Is just `#` marked as `LT_PreprocessorDirective`?

File: clang/unittests/Format/FormatTest.cpp [hunk 13]
```diff
@@ -24206,10 +24206,11 @@ TEST_F(FormatTest, IgnorePPDefinitions) {
   Style.IndentPPDirectives = FormatStyle::PPDIS_None;
   verifyNoChange("#if A\n"
                  "#define A  a\n"
-                 "#endif",
+                 "#endif\n",
```
Comment (HazardyKnusperkeks) [inline clang/unittests/Format/FormatTest.cpp:24223]: We normally don't have a trailing line break in the tests. Why did you add it?

File: clang/unittests/Format/FormatTest.cpp [hunk 14]
```diff
@@ -24153,6 +24153,113 @@ TEST_F(FormatTest, WhitespaceSensitiveMacros) {
...
+  verifyNoChange("#define A  (  args   )  =  func  (  args  )", Style);
+
+  verifyNoChange("#define A x:", Style);
+  verifyNoChange("#define A a. b", Style);
+
+  // Surrounded with formatted code.
+  verifyFormat("int a;\n"
+               "#define  A  a\n"
+               "int a;",
+               "int  a ;\n"
```
Comment (mydeveloperday) [inline clang/unittests/Format/FormatTest.cpp:24172]: Why is that not formatted

**Review by mydeveloperday — COMMENTED** [review]
Something here doesn't feel right

File: clang/unittests/Format/FormatTest.cpp (hunk 13 above)
Comment (tomekpaszek) [reply clang/unittests/Format/FormatTest.cpp]: no reason. removed the trailing line breaks now

File: clang/lib/Format/UnwrappedLineFormatter.cpp (hunk 12 above)
Comment (tomekpaszek) [reply clang/lib/Format/UnwrappedLineFormatter.cpp]: Good point, better safe than sorry. I've refactored the condition in the next changeset.

File: clang/unittests/Format/FormatTest.cpp (hunk 14 above)
Comment (tomekpaszek) [reply clang/unittests/Format/FormatTest.cpp:24172]: Not sure what are you referring to in this case

**Review by tomekpaszek — COMMENTED** [review]
>Something here doesn't feel right

Could you be more specific?

File: clang/unittests/Format/FormatTest.cpp (hunk 14 above)
Comment (tomekpaszek) [reply clang/unittests/Format/FormatTest.cpp:24172]: This practical line is the start of the code that should be formatted. The expected output starts 3 lines above

Comment (HazardyKnusperkeks) [reply clang/unittests/Format/FormatTest.cpp:24172]: It's often hard to see that comma. That's what probably happened here.

File: clang/lib/Format/UnwrappedLineFormatter.cpp [hunk 15]
```diff
@@ -1308,6 +1308,13 @@ class OptimizingLineFormatter : public LineFormatter {
 
 } // anonymous namespace
 
+static bool lineContainsPPDefinition(const AnnotatedLine &Line) {
```
Comment (HazardyKnusperkeks) [inline clang/lib/Format/UnwrappedLineFormatter.cpp:1311]:
```suggestion
static bool LineContainsPPDefinition(const AnnotatedLine &Line) {
```
Code style.

File: clang/lib/Format/UnwrappedLineFormatter.cpp [hunk 16]
```diff
@@ -1355,6 +1362,10 @@ unsigned UnwrappedLineFormatter::format(
     bool FixIndentation = (FixBadIndentation || ContinueFormatting) &&
                           Indent != TheLine.First->OriginalColumn;
     bool ShouldFormat = TheLine.Affected || FixIndentation;
+
+    if (Style.IgnorePPDefinitions && lineContainsPPDefinition(TheLine))
+      ShouldFormat = false;
```
Comment (HazardyKnusperkeks) [inline clang/lib/Format/UnwrappedLineFormatter.cpp:1367]:
```suggestion
    bool ShouldFormat = (TheLine.Affected || FixIndentation) && (!Style.IgnorePPDefinitions || lineContainsPPDefinition(TheLine));
```
Maybe? Don't know which is better.

Comment (tomekpaszek) [reply clang/lib/Format/UnwrappedLineFormatter.cpp:1367]: Personally, I like the original approach but maybe having longer conditions is more in line with the rest of the code. Let me know if I should change it.

Comment (HazardyKnusperkeks) [reply clang/lib/Format/UnwrappedLineFormatter.cpp]: I will not request the change. :)
I'm fine with either way.

**Review by HazardyKnusperkeks — APPROVED** [review]
LGTM, but wait for either of @owenca or @mydeveloperday .

File: clang/include/clang/Format/Format.h [hunk 17]
```diff
@@ -2443,6 +2443,10 @@ struct FormatStyle {
   /// \version 13
   std::vector<std::string> IfMacros;
 
+  /// Ignore formatting in preprocessor definitions.
```
Comment (owenca) [inline clang/include/clang/Format/Format.h:2446]:
```suggestion
  /// Do not format macro definitions.
```

File: clang/unittests/Format/FormatTest.cpp [hunk 18]
```diff
@@ -24153,6 +24153,123 @@ TEST_F(FormatTest, WhitespaceSensitiveMacros) {
   verifyNoChange("FOO(String-ized&Messy+But: :Still=Intentional);", Style);
 }
 
+TEST_F(FormatTest, IgnorePPDefinitions) {
+  FormatStyle Style = getLLVMStyle();
```
Comment (owenca) [inline clang/unittests/Format/FormatTest.cpp:24157]:
```suggestion
  auto Style = getLLVMStyle();
```

**Review by owenca — COMMENTED** [review]
Please update ReleaseNotes.rst.

File: clang/unittests/Format/FormatTest.cpp [hunk 19]
```diff
@@ -24153,6 +24153,123 @@ TEST_F(FormatTest, WhitespaceSensitiveMacros) {
...
+                 "#endif",
+                 Style);
+  verifyNoChange("#define UNITY 1\n"
+                 "#if A\n"
+                 "#  define   A  a\\\n"
+                 "  a  a\n"
+                 "#endif",
+                 Style);
+  verifyFormat("#if A\n"
+               "#define A  a\n"
```
Comment (owenca) [inline clang/unittests/Format/FormatTest.cpp:24232]: The indentation should _not_ change.

File: clang/unittests/Format/FormatTest.cpp [hunk 20]
```diff
@@ -24153,6 +24153,123 @@ TEST_F(FormatTest, WhitespaceSensitiveMacros) {
...
+               "#endif",
+               "#if A\n"
+               "  #define A  a\n"
+               "#endif",
+               Style);
+  Style.IndentPPDirectives = FormatStyle::PPDIS_AfterHash;
+  verifyFormat("#if A\n"
+               "#  define A  a\n"
```
Comment (owenca) [inline clang/unittests/Format/FormatTest.cpp:24240]: Ditto.

File: clang/unittests/Format/FormatTest.cpp [hunk 21]
```diff
@@ -24153,6 +24153,123 @@ TEST_F(FormatTest, WhitespaceSensitiveMacros) {
...
+  Style.IndentPPDirectives = FormatStyle::PPDIS_BeforeHash;
+  verifyNoChange("#if A\n"
+                 "  #define A  a\n"
+                 "#endif",
+                 Style);
+
+  Style.IndentPPDirectives = FormatStyle::PPDIS_None;
+  // IgnorePPDefinitions should not affect other PP directives
+  verifyFormat("#if !defined(A)\n"
+               "# define  A  a\n"
```
Comment (owenca) [inline clang/unittests/Format/FormatTest.cpp:24255]: Ditto.

File: clang/unittests/Format/FormatTest.cpp [hunk 22]
```diff
@@ -24153,6 +24153,123 @@ TEST_F(FormatTest, WhitespaceSensitiveMacros) {
...
+               "#if ! defined ( A )\n"
+               "  # define  A  a\n"
+               "#endif",
+               Style);
+
+  // With comments.
+  verifyNoChange("/* */  # define A  a  //  a  a", Style);
+  verifyFormat("int a;     // a\n"
+               "#define A  // a\n"
+               "int aaa;   // a",
```
Comment (owenca) [inline clang/unittests/Format/FormatTest.cpp:24266]: This seems inconsistent with lines 24186-24188 above.

**Review by owenca — CHANGES_REQUESTED** [review]

**Comment by owenca** [comment]
After giving more thoughts to this, I think what we really want is `SkipMacroDefinitionBody`, which would format the code below:
```
   # define   A   a. b   //comment
   # define   A( x , y )   ( ( x ) + ( y ) )
```
To the following:
```
#define A   a. b // comment
#define A(x, y)   ( ( x ) + ( y ) )
```
That is, only the body (except comments) of a macro definition is not formatted.

**Comment by tomekpaszek** [comment]
> After giving more thoughts to this, I think what we really want is `SkipMacroDefinitionBody`, which would format the code below:
> (...)
> That is, only the body (except comments) of a macro definition is not formatted.

I understand that: 
- we do want to align the macros (respect IndentPPDirectives)
- we do want to remove extra whitespaces before, within and right after the first token (#define)
- we do NOT want to touch the body
- we do NOT want to break lines

**Comment by owenca** [comment]
> > After giving more thoughts to this, I think what we really want is `SkipMacroDefinitionBody`, which would format the code below:
> > (...)
> > That is, only the body (except comments) of a macro definition is not formatted.
> 
> I understand that:
> 
> * we do want to align the macros (respect IndentPPDirectives)
> * we do want to remove extra whitespaces before, within and right after the first token (#define)

We want to format e.g. `#define FOO` and `#define BAR(x, y)`.

> * we do NOT want to touch the body
> * we do NOT want to break lines

We don't want to change line splicing (if any) within the body.

File: clang/lib/Format/UnwrappedLineParser.cpp [hunk 23]
```diff
@@ -1157,7 +1157,15 @@ void UnwrappedLineParser::parsePPDefine() {
...
   FormatTok->Tok.setIdentifierInfo(Keywords.kw_internal_ident_after_define);
-  nextToken();
+
+  if (Style.SkipMacroDefinitionBody) {
+    do {
+      nextToken();
+    } while (!eof());
+  } else {
+    nextToken();
+  }
```
Comment (owenca) [inline clang/lib/Format/UnwrappedLineParser.cpp:1167]: Instead, we can simply mark the tokens `Finalized` (after `Line->InMacroBody` is set to `true` below) so that the continuation indenter can be left alone.

**Comment by tru** [comment]
@tomekpaszek Hi Tomek, do you think you will have time to work on this soon? Otherwise, I can probably finish it off for you since we also want this feature.

**Comment by tomekpaszek** [comment]
> @tomekpaszek Hi Tomek, do you think you will have time to work on this soon? Otherwise, I can probably finish it off for you since we also want this feature.

Hi tru, 
Thanks for reaching out. Unfortunately I'm too busy to work on this at the moment. Feel free to push it forward.

**Comment by tru** [comment]
@owenca What's the things that still needs to be addressed for this to land?

File: clang/lib/Format/UnwrappedLineParser.cpp (hunk 23 above)
Comment (tomekpaszek) [reply clang/lib/Format/UnwrappedLineParser.cpp:1166]: Hi!
Maybe it's a little late, but I wanted to let you know that I've tried that but it didn't give the expected results. Didn't have time to dig into this approach.

Comment (owenca) [reply clang/lib/Format/UnwrappedLineParser.cpp:1166]: Np!
